### PR TITLE
storage: in persist_source, fix locking

### LINKED ISSUE #16090: storage: `persist_source` latency scales with number of workers
We noticed in a customer environment that the latency of queries against a persist source seems to scale with the number of workers per-process in the cluster. Here are some timings:

```
2xsmall
materialize=> select * from mz_views;
Time: 484.473 ms
materialize=> select * from mz_views;
Time: 397.870 ms
materialize=> select * from mz_views;
Time: 604.381 ms
materialize=> select * from mz_views;
Time: 412.148 ms

xsmall
materialize=> select * from mz_views;
Time: 450.758 ms
materialize=> select * from mz_views;
Time: 400.150 ms
materialize=> select * from mz_views;
Time: 393.060 ms
materialize=> select * from mz_views;
Time: 429.285 ms

small
materialize=> select * from mz_views;
Time: 587.237 ms
materialize=> select * from mz_views;
Time: 608.190 ms
materialize=> select * from mz_views;
Time: 557.064 ms
materialize=> select * from mz_views;
Time: 482.227 ms
materialize=> select * from mz_views;
Time: 541.872 ms

medium
materialize=> select * from mz_views;
Time: 1037.786 ms (00:01.038)
materialize=> select * from mz_views;
Time: 1032.083 ms (00:01.032)
materialize=> select * from mz_views;
Time: 1065.255 ms (00:01.065)
materialize=> select * from mz_views;
Time: 1315.665 ms (00:01.316)

large
materialize=> select * from mz_views;
Time: 2104.311 ms (00:02.104)
materialize=> select * from mz_views;
Time: 1607.603 ms (00:01.608)
materialize=> select * from mz_views;
Time: 1862.535 ms (00:01.863)
materialize=> select * from mz_views;
Time: 2551.174 ms (00:02.551)

xlarge
materialize=> select * from mz_views;
Time: 3973.676 ms (00:03.974)
materialize=> select * from mz_views;
Time: 4989.184 ms (00:04.989)
materialize=> select * from mz_views;
Time: 4526.773 ms (00:04.527)
materialize=> select * from mz_views;
Time: 4216.659 ms (00:04.217)

2xlarge
materialize=> select * from mz_views;
Time: 4577.487 ms (00:04.577)
materialize=> select * from mz_views;
Time: 2157.412 ms (00:02.157)
materialize=> select * from mz_views;
Time: 5688.162 ms (00:05.688)
materialize=> select * from mz_views;
Time: 4961.007 ms (00:04.961)
materialize=> select * from mz_views;
Time: 3797.427 ms (00:03.797)
```

Of particular interest is the fact that the performance seems stable in a 2xlarge. This suggests that there is some sort of per _process_ bottleneck, rather than a per _worker_ bottleneck.

Initial hypothesis, from discussion with @danhhz, is that we are serially acquiring a mutex, worker by worker, while opening persist connections. This would explain why the latency scales up with number of workers _until_ you add another process, at which point the latency stays constant, as each process gets a fresh mutex.

This seems to reproduce readily in a fresh cloud environment. It may also reproduce locally—no one's tried yet!

@aljoscha @danhhz or @uce—could you prioritize this one?

### PR DESCRIPTION
**Pull request by aljoscha**
### Motivation

Fixes #16090

### Tips for reviewer

The first commit (fixing locking) does mitigate things, but the bad scaling is still easy to observe locally. The second commit (creating a `BatchFetcher` without going through a `ReadHandle`) makes the bad scaling almost unnoticeable but seems more controversial because it adds a new method to `PersistClient` that seems easy to mis-use.

Btw, I reproduced this locally by just creating a 1-worker cluster and a 32-worker cluster:
```
materialize=> CREATE CLUSTER c32 REPLICAS (r1 (SIZE '32'));
materialize=> CREATE CLUSTER c1 REPLICAS (r1 (SIZE '1'));
```
switching between them, and observing the runtime of `SHOW VIEWS`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by aljoscha** [comment]
Some more colour for what the second fix does: before this change, all workers are trying to concurrently register a reader in shard state, meaning there will be lots of conflicts. Eventually they all succeed, and then create the `BatchFetcher`, which destroys the reader in in shard state again. Just not doing that improves startup time by quite a bit.

File: src/storage-client/src/source/persist_source.rs [hunk 1]
```diff
@@ -250,11 +250,14 @@ where
...
-            .open(persist_location_stream)
-            .await;
+        let read = {
+            let mut persist_clients = persist_clients_stream.lock().await;
+
+            let read = persist_clients.open(persist_location_stream).await;
+            std::mem::drop(persist_clients);
+
+            read
+        };
```
Comment (guswynn) [inline src/storage-client/src/source/persist_source.rs:260]:
```suggestion
        let read = {
            let mut persist_clients = persist_clients_stream.lock().await;
            persist_clients.open(persist_location_stream).await;
        };
```

I think is more clear! but this is a nit

File: src/storage-client/src/source/persist_source.rs [hunk 2]
```diff
@@ -414,17 +417,21 @@ where
...
-            .await;
+        let fetcher = {
+            let mut persist_clients = persist_clients.lock().await;
+
+            let persist_client = persist_clients
+                .open(metadata.persist_location.clone())
+                .await
+                .expect("could not open persist client");
+
+            std::mem::drop(persist_clients);
```
Comment (guswynn) [inline src/storage-client/src/source/persist_source.rs:428]:
```suggestion
            // unlock the client cache before we do any async work
            std::mem::drop(persist_clients);
```

File: src/persist-client/src/lib.rs [hunk 3]
```diff
@@ -453,6 +456,25 @@ impl PersistClient {
         Ok(reader)
     }
 
+    /// Creates and returns a [BatchFetcher] for the given shard id.
+    #[instrument(level = "debug", skip_all, fields(shard = %shard_id))]
+    pub fn create_batch_fetcher<K, V, T, D>(&self, shard_id: ShardId) -> BatchFetcher<K, V, T, D>
```
Comment (guswynn) [inline src/persist-client/src/lib.rs:461]:
```suggestion
    /// Creates and returns a [BatchFetcher] for the given shard id. This method can be misused, and requires
    /// that you ensure XXXX.
    #[instrument(level = "debug", skip_all, fields(shard = %shard_id))]
    pub fn unchecked_create_batch_fetcher<K, V, T, D>(&self, shard_id: ShardId) -> BatchFetcher<K, V, T, D>
```

I think its fine to add methods that increase perf at the cost of misuse, as long as they are appropriately named and documented

File: src/persist-client/src/lib.rs [hunk 4]
```diff
@@ -453,6 +456,25 @@ impl PersistClient {
...
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64 + Send + Sync,
+    {
+        let fetcher = BatchFetcher {
```
Comment (danhhz) [inline src/persist-client/src/lib.rs:468]: how would we feel about something like calling `StateVersions::fetch_current_state` or `StateVersions::maybe_init_shard` here as a mechanism to detect misuse. that would read from the network, but it wouldn't serialize on anything (or write to state)

**Review by danhhz — APPROVED** [review]
LGTM but strong preference for doing the one comment I left

**Comment by danhhz** [comment]
Also fyi @petrosagg mentioned recently that he's on board with pushing the Mutex into PersistClientCache, which I think would eliminate this entire class of mutex guard problems

File: src/persist-client/src/lib.rs (hunk 4 above)
Comment (danhhz) [reply src/persist-client/src/lib.rs:468]: to ensure the codecs match

**Comment by benesch** [comment]
So good to have you back, @aljoscha! Thanks for turning this one around so quickly! 

Is there an easy way to lock this perf improvement in with a test? Are the feature benchmarks running these days? cc @philip-stoev

**Comment by philip-stoev** [comment]
They are not running right now unfortunately. If this PR will remain open for the next 2 days, I can try to revive them. Please let me know.

File: src/persist-client/src/lib.rs (hunk 4 above)
Comment (aljoscha) [reply src/persist-client/src/lib.rs:488]: @guswynn I added this, as a check, so we don't need to rename to `unchecked_` and add the safety warning. Would you agree?

**Comment by aljoscha** [comment]
Also confirmed the fix in staging: without the fix a `SHOW INDEXES` on an `xlarge` cluster takes ~66 seconds, with the fix, it takes ~600 ms. Yes, the first one is in seconds!

**Comment by aljoscha** [comment]
@philip-stoev & @benesch Do we want to keep the PR open this long? I might even be inclined to advocate for including this in today's release even though it's past the cutoff window: the fix is simple enough, and it's a big improvement in usability (see that 60s vs 600ms number above). WDYT?

**Comment by philip-stoev** [comment]
We , and by we I mean I, am not set up to do same-day signoffs on stuff part the release cut-off time. I am going to schedule a Nightly run, this is the best I can do on such a short notice .

**Comment by danhhz** [comment]
I'm not particularly worried about this having unintended side effects. One of the commits is just reducing the static scope that a MutexGuard exists for and Rust's checking is enough to ensure the new shorter lifetime is safe. The other was a known TODO from when the code in question first went in.

So, I'd be comfortable merging this and cherry-picking it into v0.32.1 if we think the customer impact is worth it.

**Pull request merged**